commit 92d9ab4c28842a09ca2b76d3ff2f933e01b6cd6f
Author: Nick Wellnhofer <wellnhofer@aevum.de>
Date:   Mon Jun 7 15:09:53 2021 +0200

    Fix whitespace when serializing empty HTML documents
    
    The old, non-recursive HTML serialization code would always terminate
    the output with a newline. The new implementation omitted the newline
    if the document node had no children. Readd the newline when
    serializing empty documents.
    
    Fixes #266.

diff --git a/HTMLtree.c b/HTMLtree.c
index bdd639c7..7a2b8558 100644
--- a/HTMLtree.c
+++ b/HTMLtree.c
@@ -763,11 +763,15 @@ htmlNodeDumpFormatOutput(xmlOutputBufferPtr buf, xmlDocPtr doc,
             if (((xmlDocPtr) cur)->intSubset != NULL) {
                 htmlDtdDumpOutput(buf, (xmlDocPtr) cur, NULL);
             }
-            /* Always validate cur->parent when descending. */
-            if ((cur->parent == parent) && (cur->children != NULL)) {
-                parent = cur;
-                cur = cur->children;
-                continue;
+            if (cur->children != NULL) {
+                /* Always validate cur->parent when descending. */
+                if (cur->parent == parent) {
+                    parent = cur;
+                    cur = cur->children;
+                    continue;
+                }
+            } else {
+                xmlOutputBufferWriteString(buf, "\n");
             }
             break;
 
